File: CFPT_2025_RESTAT_main.R
###############################################################################
## Replication File for Cattaneo, Feng, Palomba, and Titiunik (2025)
##  
## This code replicates the main analysis of the paper and the robustness
## exercises in the supplemental appendix.
###############################################################################

rm(list=ls(all=TRUE))

##########################################
# Load stuff 
pacman::p_load(scpi, haven, wesanderson, ggplot2, reshape2, devtools, tidyverse)
devtools::install_github("apoorvalal/LalRUtils")

##########################################
# Set paths
path <- "YOUR_PATH"
cores <- 1

path.data <- paste0(path, "data/")
path.fig  <- paste0(path, "fig/")
path.out  <- paste0(path, "out/")
path.code <- paste0(path, "code/")
source(paste0(path.code, "0001_funs.R"))

##########################################
# Set parameters/lists

arab.league <- c("Algeria", "Egypt", "Libya", "Morocco", "Sudan", 
                 "Tunisia", "Djibouti", "Mauritania", "Somalia")
sims <- 200L
post.est <- 5L
anticipation <- 0L
covs.adj <- features <- NULL
coig.data <- TRUE
constant <- TRUE


##########################################
# Load data
##########################################

data <- haven::read_dta(paste0(path.data, "final_data.dta"))
data$lgdp <- log(data$rgdppp)

# remove countries in the arab league
data <- subset(data, continent == "Africa" & !(countryname %in% arab.league))

##########################################
## 1) Main Empirical Analysis
##########################################

set.seed(8894)

# We consider four different groups of treated units
treated.units.list <- list("WaveAll" = unique(subset(data, treated == 1 & trDate <= 1994)$countryname),
                           "Wave1" = unique(subset(data, treated == 1 & trDate < 1987)$countryname),
                           "Wave2" = unique(subset(data, treated == 1 & trDate >= 1987 & trDate <= 1991)$countryname),
                           "Wave3" = unique(subset(data, treated == 1 & trDate > 1991 & trDate <= 1994)$countryname))

# and three different methodologies
methods.list <- c("ridge", "simplex", "L1-L2")

##########################################
## 1a) one feature (log rgdp per capita)

for (method in methods.list) {
  for (j in seq_len(length(treated.units.list))) {
    treated.units <- treated.units.list[[j]]
    wave.name <- names(treated.units.list)[j]
    save.name <- paste0("Africa_", wave.name, "_", method)
    
    if (wave.name=="WaveAll") {
      eff.it <- eff.i <- eff.t <- TRUE
    } else {
      eff.it <- eff.i <- FALSE; eff.t <- TRUE
    }
    
    # save .RData 
    predictandsGet(data, treated.units, w.constr=list("name" = method), 
                   save.path=path.out, save.name=save.name,
                   anticipation=anticipation, post.est=post.est, sims=sims,
                   constant=TRUE, cores=cores, 
                   eff.it=eff.it, eff.i=eff.i, eff.t=eff.t)
  }
}

##############################################################
## 1b) two features (log rgdp per capita, investment over gdp)

features <- list(c("lgdp", "inv_ratio"))
covs.adj <- list("lgdp"= c("constant", "trend"), 
                 "inv_ratio"= c("constant", "trend"))

for (method in methods.list) {
  for (j in seq_len(length(treated.units.list))) {

    treated.units <- treated.units.list[[j]]
    wave.name <- names(treated.units.list)[j]
    save.name <- paste0("Africa_", wave.name, "_", method, "_covs")
 
    if (wave.name=="WaveAll") {
      eff.it <- eff.i <- eff.t <- TRUE
    } else {
      eff.it <- eff.i <- FALSE; eff.t <- TRUE
    }    
     
    # save .RData 
    predictandsGetCovs(data, treated.units, w.constr=list("name" = method), 
                       save.path=path.out, save.name=save.name,
                       features=features, covs.adj=covs.adj, 
                       anticipation=anticipation, post.est=post.est, sims=sims,
                       constant=TRUE, cores=cores, 
                       eff.it=eff.it, eff.i=eff.i, eff.t=eff.t)
  }
}


##########################################
## 2) Robustness Exercises 
##########################################

##########################################
## 2a) Placebo date

Pl.delta <- 11
data$trDate <- data$trDate - Pl.delta
data$liberalization <- 1*(data$year >= data$trDate)
data <- subset(data, countryname != "Mauritius") # Mauritius just has 5 pre-treatment periods

treated.units.list <- list("WaveAll" = unique(subset(data, treated == 1 & trDate <= (1994 - Pl.delta))$countryname),
                           "Wave1" = unique(subset(data, treated == 1 & trDate < (1987 - Pl.delta))$countryname),
                           "Wave2" = unique(subset(data, treated == 1 & trDate >= (1987 - Pl.delta) & trDate <= (1991 - Pl.delta))$countryname),
                           "Wave3" = unique(subset(data, treated == 1 & trDate > (1991-Pl.delta) & trDate <= (1994-Pl.delta))$countryname))

method <- "L1-L2"

for (j in seq_len(length(treated.units.list))) {
  treated.units <- treated.units.list[[j]]
  wave.name <- names(treated.units.list)[j]
  save.name <- paste0("Africa_", wave.name, "_", method, "_placebo")
  
  if (wave.name=="WaveAll") {
    eff.it <- eff.i <- eff.t <- TRUE
  } else {
    eff.it <- eff.i <- FALSE; eff.t <- TRUE
  }
  
  # save .RData 
  predictandsGet(data, treated.units, w.constr=list("name" = method), 
                 save.path=path.out, save.name=save.name,
                 anticipation=anticipation, post.est=post.est, sims=sims,
                 constant=TRUE, cores=cores, 
                 eff.it=eff.it, eff.i=eff.i, eff.t=eff.t, placebo.time=TRUE)
}

File: CFPT_2025_RESTAT_main.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Dec 23 16:28:10 2024

@author: fpalomba
"""

###################################################################
# Replication file for Cattaneo, Feng, Palomba, and Titiunik (2025)
# This code replicates the main analysis of the paper using Python
# and it serves for illustrative purposes only. All the robustness
# exercises are conducted only in R.
###################################################################

########################################
# Load modules
import pandas
import numpy
import os
from warnings import filterwarnings
from plotnine import theme

from scpi_pkg.scdataMulti import scdataMulti
from scpi_pkg.scpi import scpi
from scpi_pkg.scplotMulti import scplotMulti

# set seed, paths, number of simulations, and cores
filterwarnings("ignore")
numpy.random.seed(8894)

os.chdir('YOUR_PATH_HERE')

sims = 200
cores = 1

########################################
# Load database
data = pandas.read_stata('data/BNdata.dta')
data['lgdp'] = numpy.log(data['rgdppp'])
data['year'] = pandas.DatetimeIndex(data['year']).year  # from datetime64 to int

# remove countries not in africa and not in the arab league
data = data.loc[data['continent'] == 'Africa']
arab_league = ["Algeria", "Egypt", "Libya", "Morocco", "Sudan",
               "Tunisia", "Djibouti", "Mauritania", "Somalia"]
data = data[~data['countryname'].isin(arab_league)]

# prepare four main sets of treated units
WaveAll = data.query("treated == 1 and trDate <= 1994")["countryname"].unique().tolist()
Wave1 = data.query("treated == 1 and trDate < 1987")["countryname"].unique().tolist()
Wave2 = data.query("treated == 1 and 1987 <= trDate <= 1991")["countryname"].unique().tolist()
Wave3 = data.query("treated == 1 and 1991 < trDate <= 1994")["countryname"].unique().tolist()

units_dict = {'WaveAll': WaveAll,
              'Wave1': Wave1,
              'Wave2': Wave2,
              'Wave3': Wave3
              }

###############################################
# TSUS \tau_{ik} - All countries
###############################################
aux = scdataMulti(df=data,
                  id_var='countryname',
                  treatment_var='liberalization',
                  outcome_var='lgdp',
                  time_var='year',
                  constant=True,
                  cointegrated_data=True,
                  post_est=5,
                  units_est=units_dict["WaveAll"],
                  effect='unit-time')

res = scpi(aux, w_constr={'name': 'L1-L2'}, sims=sims,
           u_order=1, u_lags=1, cores=cores)
p = scplotMulti(res, ptype='series', scales='free', ncols=4)
p = p + theme(legend_position='bottom')
plotname = 'fig/other/Africa_WaveAll_indiv_L1-L2_py.png'
p.save(plotname)

###############################################
# TAUS \tau_{i.} - All countries
###############################################
aux = scdataMulti(df=data,
                  id_var='countryname',
                  treatment_var='liberalization',
                  outcome_var='lgdp',
                  time_var='year',
                  constant=True,
                  cointegrated_data=True,
                  post_est=5,
                  units_est=units_dict["WaveAll"],
                  effect='unit')

res = scpi(aux, w_constr={'name': 'L1-L2'}, sims=sims,
           u_order=1, u_lags=1, cores=cores)
p = scplotMulti(res, ptype='series', scales='free', ncols=4)
p = p + theme(legend_position='bottom')
plotname = 'fig/other/Africa_WaveAll_unit_L1-L2_py.png'
p.save(plotname)

###############################################
# TSUA \tau_{.t} - All countries
###############################################
treated_units_time = [tr for tr in units_dict["WaveAll"] if tr != "Mauritius"]

aux = scdataMulti(df=data,
                  id_var='countryname',
                  treatment_var='liberalization',
                  outcome_var='lgdp',
                  time_var='year',
                  constant=True,
                  cointegrated_data=True,
                  post_est=5,
                  units_est=treated_units_time,
                  effect='time')

res = scpi(aux, w_constr={'name': 'L1-L2'}, sims=sims,
           u_order=1, u_lags=1, cores=cores)
p = scplotMulti(res, ptype='series', scales='free', joint=True)
p = p + theme(legend_position='bottom')
plotname = 'fig/other/Africa_WaveAll_time_L1-L2_py.png'
p.save(plotname)

###############################################
# TSUA \tau_{.t} - Before 1991
###############################################
treated_units_time = [tr for tr in units_dict["Wave1"] if tr != "Mauritius"]

aux = scdataMulti(df=data,
                  id_var='countryname',
                  treatment_var='liberalization',
                  outcome_var='lgdp',
                  time_var='year',
                  constant=True,
                  cointegrated_data=True,
                  post_est=5,
                  units_est=treated_units_time,
                  effect='time')

res = scpi(aux, w_constr={'name': 'L1-L2'}, sims=sims,
           u_order=1, u_lags=1, cores=cores)
p = scplotMulti(res, ptype='series', scales='free', joint=True)
p = p + theme(legend_position='bottom')
plotname = 'fig/other/Africa_Wave1_time_L1-L2_py.png'
p.save(plotname)

###############################################
# TSUA \tau_{.t} - Between 1987 and 1991
###############################################
aux = scdataMulti(df=data,
                  id_var='countryname',
                  treatment_var='liberalization',
                  outcome_var='lgdp',
                  time_var='year',
                  constant=True,
                  cointegrated_data=True,
                  post_est=5,
                  units_est=units_dict["Wave2"],
                  effect='time')

res = scpi(aux, w_constr={'name': 'L1-L2'}, sims=sims,
           u_order=1, u_lags=1, cores=cores)
p = scplotMulti(res, ptype='series', scales='free', joint=True)
p = p + theme(legend_position='bottom')
plotname = 'fig/other/Africa_Wave2_time_L1-L2_py.png'
p.save(plotname)

###############################################
# TSUA \tau_{.t} - After 1991
###############################################
aux = scdataMulti(df=data,
                  id_var='countryname',
                  treatment_var='liberalization',
                  outcome_var='lgdp',
                  time_var='year',
                  constant=True,
                  cointegrated_data=True,
                  post_est=5,
                  units_est=units_dict["Wave3"],
                  effect='time')

res = scpi(aux, w_constr={'name': 'L1-L2'}, sims=sims,
           u_order=1, u_lags=1, cores=cores)
p = scplotMulti(res, ptype='series', scales='free', joint=True)
p = p + theme(legend_position='bottom')
plotname = 'fig/other/Africa_Wave3_time_L1-L2_py.png'
p.save(plotname)
